spacegrep/src/bin/Spacegrep_main.ml

```diff
@@ -43,9 +43,22 @@ let run_all
     ~case_sensitive ~debug ~force ~output_format ~highlight ~warn
     patterns docs =
   let matches =
-    List.map (fun get_doc_src ->
-      let doc_src = get_doc_src () in
-      let doc_type = File_type.classify doc_src in
+    List.filter_map (fun (get_doc_src : ?max_len:int -> unit -> Src_file.t) ->
+      (*
+         We inspect the first 4096 bytes to guess whether the file type.
+         This saves time on large files, by reading typically just one
+         block from the file system.
+      *)
+      let peek_length = 4096 in
+      let partial_doc_src = get_doc_src ~max_len:peek_length () in
+      let doc_type = File_type.classify partial_doc_src in
+      let doc_src =
```

Comment:
Shoudn't this be put later when you know you're not in the Minified or Binary case? Otherwise
it defeats the purpose of the optimization no? Looks like you gonna read the whole doc in every paths,
whatever the result of the classify on the peek_length read doc.

Reply:
ow, I really rushed this. Thanks for taking a look.

Reply:
weird that you got better performance ... it should have been the same amount of time than before.

Reply:
I fixed this, which seems to improves the performance overall and on a single big binary file, but otherwise it's slower on a repo with mostly small source files. I don't understand why. I'll try to get to the bottom of this before merging. I might also postpone this task since it's fun but not urgent.